### PR TITLE
Implement Recipe API

### PR DESCRIPTION
**Pull request by Faithcaio**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1582)|*SpongeCommon*|[SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1593)

Continuation of https://github.com/SpongePowered/SpongeCommon/pull/1119 of https://github.com/SpongePowered/SpongeAPI/pull/1098
Updated for MC 1.12
___

`InventoryUtil`: CraftingInventory <-> InventoryCrafting Util
`ItemStackUtil`: List<Ingredient> <-> native Ingredient[]

`AbstractSpongeCraftingRecipe`/`DelegateSpongeCraftingRecipe`: for custom CraftingRecipe Impls.
`CustomIngredient`: For Ingredients using Predicates instead of Items.
`DelegateIngredient`: For custom Ingredient Impls.
`IngredientUtil`: Ingredient <-> native Ingredient
`SpongeCraftingRecipeRegistry`: delegates registration to the `CraftingManager`
`SpongeIngredientBuilder`: Builder for Ingredients
`SpongeShapedCraftingRecipeBuilder`: Builder for shaped CraftingRecipes
`SpongeShapelessCraftingRecipeBuilder`: Builder for shapeless CraftingRecipes

`MatchSmeltingVanillaItemStack`:
`SpongeSmeltingRecipe`:
`SpongeSmeltingRecipeBuilder`

`MixinContainerCrafting`:
`MixinIRecipe`: make IRecipe implement CraftingRecipe
`MixinIngredient`: make native Ingredient implement Ingredient
`MixinShapedRecipes`: make ShapedRecipes implement ShapedCraftingRecipe
`MixinShapelessRecipes`: make ShapelessRecipes implement ShapelessCraftingRecipe
`MixinFurnaceRecipes`: make FurnaceRecipes implement SmeltingRecipeRegistry

`CommonModuleRegistry`: register Builders: `Ingredient`, `ShapedCraftingRecipe`, `ShapelessCraftingRecipe`, `SmeltingRecipe`

___

Ideas for Further enhancement:
* Pass the crafter to the recipe to allow for more customization. Possible?

### CODE REVIEW

File: src/main/java/org/spongepowered/common/item/recipe/crafting/CustomIngredient.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        super(ItemStackUtil.toNative(matchingStacks));
+        this.predicates = predicates;
+        this.items = items;
+    }
+
+    @Override
+    public boolean apply(@Nullable net.minecraft.item.ItemStack p_apply_1_) {
+
+        // first check for matching predicates
+        if (predicates.stream().anyMatch(p -> p.test(ItemStackUtil.fromNative(p_apply_1_)))) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/CustomIngredient.java]: `this.`

File: src/main/java/org/spongepowered/common/item/recipe/crafting/CustomIngredient.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    private final List<ItemStack> items;
+
+    public CustomIngredient(List<Predicate<ItemStack>> predicates, List<ItemStack> items, List<ItemStack> matchingStacks) {
+        super(ItemStackUtil.toNative(matchingStacks));
+        this.predicates = predicates;
+        this.items = items;
+    }
+
+    @Override
+    public boolean apply(@Nullable net.minecraft.item.ItemStack p_apply_1_) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/CustomIngredient.java]: parameter could use a better name.

File: src/main/java/org/spongepowered/common/item/recipe/crafting/DelegateIngredient.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    private org.spongepowered.api.item.recipe.crafting.Ingredient delegate;
+
+    private DelegateIngredient(org.spongepowered.api.item.recipe.crafting.Ingredient delegate) {
+        super(ItemStackUtil.toNative(delegate.ingredients()));
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean apply(@Nullable ItemStack p_apply_1_) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/DelegateIngredient.java]: parameter could use a better name

File: src/main/java/org/spongepowered/common/item/recipe/crafting/DelegateIngredient.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return delegate.test(ItemStackUtil.fromNative(p_apply_1_));
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/DelegateIngredient.java]: `this.`

File: src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeCraftingRecipeRegistry.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public void remove(CraftingRecipe recipe) {
+        // TODO how to remove from RegistryNamespaced?
+    }
+
+    @Override
+    public Collection<CraftingRecipe> getRecipes() {
+        // TODO Unmodifiable?
+        return Streams.stream(CraftingManager.REGISTRY.iterator()).map(CraftingRecipe.class::cast).collect(Collectors.toList());
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeCraftingRecipeRegistry.java]: `.collect(ImmutableList.toImmutableList())`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+public final class SpongeShapedCraftingRecipeBuilder implements ShapedCraftingRecipe.Builder {
+
+    private final List<String> aisle = Lists.newArrayList();
+    private final Map<Character, Ingredient> ingredientMap = new Char2ObjectArrayMap<>();
+    private ItemStack result = ItemStack.empty();
+    private String groupName = "";
+
+    @Override
+    public ShapedCraftingRecipe.Builder group(@Nullable String name) {
+        this.groupName = name == null ? "" : name;
```
Comment (liach) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java]: Use `Strings.nullToEmpty`

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java [hunk 7]
```diff
@@ -82,6 +83,11 @@
     @Shadow
     public abstract Slot getSlot(int slotId);
 
+    @Shadow
+    public ItemStack slotClick(int slotId, int dragType, ClickType clickTypeIn, EntityPlayer player) {
+        throw new IllegalStateException("Shadowed.");
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java:88]: Why not making this abstract?

File: src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java [hunk 8]
```diff
@@ -427,9 +431,15 @@ public Favicon loadFavicon(BufferedImage image) throws IOException {
         return SpongeFavicon.load(image);
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public RecipeRegistry getRecipeRegistry() {
-        throw new UnsupportedOperationException(); // TODO
+    public CraftingRecipeRegistry getCraftingRecipeRegistry() {
+        return new SpongeCraftingRecipeRegistry(); // TODO cache it somewhere
```
Comment (liach) [inline src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java]: Cache it in `SpongeCraftingRecipeRegistry`?

File: src/main/resources/META-INF/common_at.cfg [hunk 9]
```diff
@@ -496,3 +496,17 @@ public-f org.spongepowered.api.world.gen.type.MushroomTypes *
 public-f org.spongepowered.api.world.schematic.BlockPaletteTypes *
 
 public-f org.spongepowered.api.world.weather.Weathers *
+
+
+public net.minecraft.entity.passive.EntityVillager field_175561_bA # DEFAULT_TRADE_LIST_MAP
+public net.minecraft.entity.EntityList func_191303_a(ILjava/lang/String;Ljava/lang/Class;Ljava/lang/String;)V # register
```
Comment (liach) [inline src/main/resources/META-INF/common_at.cfg]: Are these intentional?

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java (hunk 7 above)
Comment (Faithcaio) [reply src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java:88]: MixinContainerCrafting extends this and calls super.slotClick

File: src/main/resources/META-INF/common_at.cfg (hunk 9 above)
Comment (Faithcaio) [reply src/main/resources/META-INF/common_at.cfg]: must have happened when rebasing stuff

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java (hunk 7 above)
Comment (liach) [reply src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java:88]: Oh, one of the rare cases!

File: src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeCraftingRecipeRegistry.java [hunk 10]
```diff
@@ -28,6 +28,7 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Streams;
+import com.google.inject.Singleton;
```
Comment (liach) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeCraftingRecipeRegistry.java]: Do you want this annotation?

Comment (Faithcaio) [reply src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeCraftingRecipeRegistry.java]: nobody can know!

**Review by liach — APPROVED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/smelting/MixinFurnaceRecipes.java [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    private final Map<ItemStack, SmeltingRecipe> nativeIngredientToCustomRecipe = new IdentityHashMap<>();
+
+    @Shadow public abstract ItemStack getSmeltingResult(ItemStack stack);
+    @Shadow public abstract float getSmeltingExperience(ItemStack stack);
+    @Shadow private boolean compareItemStacks(ItemStack stack1, ItemStack stack2) {
+        throw new IllegalStateException("unreachable");
+    }
+
+    @Inject(method = "getSmeltingResult", at = @At("RETURN"), cancellable = true)
+    public void onGetSmeltingResult(ItemStack stack, CallbackInfoReturnable<ItemStack> cir) {
```
Comment (stephan-gh) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/smelting/MixinFurnaceRecipes.java]: There is no need to have this public

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/smelting/MixinFurnaceRecipes.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+        if (result.isPresent()) {
+            ItemStack nativeResult = ItemStackUtil.fromSnapshotToNative(result.get().getResult());
+
+            cir.setReturnValue(nativeResult);
+        }
+    }
+
+    @Inject(method = "getSmeltingExperience", at = @At("RETURN"), cancellable = true)
+    public void onGetSmeltingExperience(ItemStack stack, CallbackInfoReturnable<Float> cir) {
```
Comment (stephan-gh) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/smelting/MixinFurnaceRecipes.java]: There is no need to have this public

**Review by stephan-gh — CHANGES_REQUESTED** [review]

**Comment by Faithcaio** [comment]
https://github.com/SpongePowered/SpongeCommon/commit/2a7a3664f2acbb92c1a5750ad288b1e242986e58

**Pull request closed without merging**